domains: port fix abort on uncaught to v0.12

whitlockjc opened the pull request:
Port fbff705 and caeb677 from v0.10 to v0.12, original commit messages:

  fbff705
  Add v8::Isolate::SetAbortOnUncaughtException() so the user can be
  notified when an uncaught exception has bubbled.

  caeb677
  Do not abort the process if an error is thrown from within a domain,
  an error handler is setup for the domain and
  --abort-on-uncaught-exception was passed on the command line.

  However, if an error is thrown from within the top-level domain's
  error handler and --abort-on-uncaught-exception was passed on the
  command line, make the process abort.

Fixes: #8877

misterdjules commented:
Thank you @whitlockjc! Tests running on [UNIX](http://jenkins.nodejs.org/job/node-test-commit-unix/150/) and on [Windows](http://jenkins.nodejs.org/job/node-test-commit-windows/148/).

misterdjules commented:
@orangemocha I've re-enabled the node-accept-pull-request job on jenkins.nodejs.org to test this PR because when I tried to use the node-test-pull-request job on jenkins-iojs.nodesource.com, it failed: https://jenkins-iojs.nodesource.com/job/node-test-pull-request/63/.

misterdjules commented:
/cc @joyent/node-collaborators

misterdjules commented:
@whitlockjc `test-domain-with-abort-on-uncaught-exception.js` is failing on [Linux](http://jenkins.nodejs.org/job/node-test-commit-unix/150/DESTCPU=ia32,label=linux/tapTestReport/simple.tap-173/) and [SmartOS](http://jenkins.nodejs.org/job/node-test-commit-unix/150/DESTCPU=ia32,label=smartos/tapTestReport/simple.tap-173/).

Fixing them should be straightforward, could you please update this PR with a fix when you have some time? Thank you!

whitlockjc commented:
You got it.  I'm sure their failures are similar to the fix for OS X where we have to check for different exit codes.  I'll update the PR.

orangemocha commented:
> @orangemocha I've re-enabled the node-accept-pull-request job on jenkins.nodejs.org to test this PR because when I tried to use the node-test-pull-request job on jenkins-iojs.nodesource.com, it failed: https://jenkins-iojs.nodesource.com/job/node-test-pull-request/63/.

The issue should be fixed: https://jenkins-iojs.nodesource.com/job/node-test-pull-request/65/

In any case, you can use node-accept-pull-request from the new Jenkins. I have re-disabled the old one now. Let me know if you still have problems. Thanks!

misterdjules commented:
A fix for similar issue(s) was merged in io.js a while ago: https://github.com/nodejs/io.js/pull/922. That fix has the advantage of not requiring to make any change to V8, so it would be great if we could use it instead of the changes in this PR.

However, it seems that that fix does not fix https://github.com/joyent/node/issues/8630:
```
➜  v0.12 git:(v0.12) ✗ cat ~/dev/repros/domain-top-level-error-handler-throw.js
var domainErrHandlerExMessage = 'exception from domain error handler';
var internalExMessage = 'You should NOT see me';

var domain = require('domain');
var d = domain.create();

d.on('error', function() {
    console.log('in domain error handler');
    throw new Error(domainErrHandlerExMessage);
});

d.run(function doStuff() {
    process.nextTick(function () {
      throw new Error(internalExMessage);
    });
});
➜  v0.12 git:(v0.12) ✗ ./node ~/dev/repros/domain-top-level-error-handler-throw.js
in domain error handler
/Users/JulienGilli/dev/repros/domain-top-level-error-handler-throw.js:14
      throw new Error(internalExMessage);
            ^
Error: You should NOT see me
    at /Users/JulienGilli/dev/repros/domain-top-level-error-handler-throw.js:14:13
    at process._tickDomainCallback (node.js:389:11)
    at Function.Module.runMain (module.js:503:11)
    at startup (node.js:129:16)
    at node.js:822:3
➜  v0.12 git:(v0.12) ✗
```

It seems to also make a node process not abort when it seems it should:

```
➜  v0.12 git:(v0.12) ✗ cat ~/dev/repros/domain-abort-on-uncaught.js
 var domain = require('domain');
var d = domain.create();
var triggeredProcessUncaughtException = false;

process.on('uncaughtException', function onUncaughtException() {
  console.log('In uncaughtException handler');
});

d.on('error', function() {
  console.log('Throwing from top level domain error handler');
  throw new Error('Error from domain error handler');
});

d.run(function doStuff() {
  // Throwing from within different types of callbacks as each of them
  // handles domains differently
  process.nextTick(function () {
    throw new Error("Error from nextTick callback");
  });

  var fs = require('fs');
  fs.exists('/non/existing/file', function onExists(exists) {
    throw new Error("Error from fs.exists callback");
  });

  setImmediate(function onSetImmediate() {
    throw new Error("Error from setImmediate callback");
  });

  throw new Error("Error from domain.run callback");
});
➜  v0.12 git:(v0.12) ✗ ./node --abort-on-uncaught-exception ~/dev/repros/domain-abort-on-uncaught.js
Throwing from top level domain error handler
In uncaughtException handler
Throwing from top level domain error handler
In uncaughtException handler
Throwing from top level domain error handler
Throwing from top level domain error handler
In uncaughtException handler
Throwing from top level domain error handler
Throwing from top level domain error handler
Throwing from top level domain error handler
In uncaughtException handler
➜  v0.12 git:(v0.12) ✗ 
```
In the test above, I would expect node to abort, not the `uncaughtException` handler to be called. But I may misinterpret the semantics of `process.on('uncaughtException')` and `--abort-on-uncaught-exception`.

@chrisdickinson Thoughts?

whitlockjc commented:
I have updated the tests to pass on Linux and SmartOS.  This involved always checking the exit signal, even though SmartOS is the only one right now that gets a signal other than `null`, and updating Linux to have variable exit codes.  I also updated the comments explaining why we use arrays of possible values for exit codes/signals.

misterdjules commented:
Thank you @whitlockjc!

Reran the whole tests suite here: https://jenkins-iojs.nodesource.com/job/node-accept-pull-request/. We have [one failure on Windows](https://jenkins-iojs.nodesource.com/job/node-test-commit-windows/nodes=win2008r2/134/tapTestReport/test.tap-173/). This is also related to changes in the implementation of `OS::Abort` in V8 from v0.10 to v0.12.

whitlockjc commented:
I'll get on it.

whitlockjc commented:
Fixed, @misterdjules was right...this was a result of `OS::Abort` changing its exit code so we now need to check for multiple values like we do on *nix.  I also cleaned up a few places in the test where lines were greater than 80 characters.

misterdjules commented:
@whitlockjc Thank you! [All tests passed](https://jenkins-iojs.nodesource.com/job/node-test-commit/163/).

whitlockjc commented:
WOOHOO!!!

jasnell commented:
@misterdjules ... is this ready to land?

whitlockjc commented:
I don't want this to drop through the cracks so can we update this with the next steps?  Really hoping to either find out that this isn't necessary in the new node repository or be told what the next steps are.  If it's still useful/important, I'll gladly port this to the new repository via a PR.

/cc @nodejs/collaborators and @nodejs/tsc

misterdjules commented:
@whitlockjc I've ported this PR in nodejs/node's master with https://github.com/nodejs/node/pull/3036 because of https://github.com/nodejs/node/issues/3035.

The next step for this PR is for me to give it a final review and land it in v0.12.

whitlockjc commented:
Thanks a lot.  I was planning on porting this today but since you beat me to it...  ;)

misterdjules commented on deps/v8/src/isolate.cc:
```diff
@@ -1090,19 +1090,30 @@ void Isolate::DoThrow(Object* exception, MessageLocation* location) {
+      // present).
       if (fatal_exception_depth == 0 &&
-          FLAG_abort_on_uncaught_exception &&
+          (FLAG_abort_on_uncaught_exception) &&
```
There should be no parentheses around `FLAG_abort_on_uncaught_exception`.

whitlockjc replied:
Yup, that was an oversight...probably a relic from previous debugging session.

misterdjules commented on src/node.cc:
```diff
@@ -909,6 +909,33 @@ Local<Value> WinapiErrnoException(Isolate* isolate,
+  }
+
+  Local<Array> domain_array = env->domain_array().As<Array>();
+  uint32_t domainsArrayLength = domain_array->Length();
```
Use snake case for consistency for `domainsArrayLength` -> `domains_array_length`, or just get rid of that temporary, it seems unneeded.

whitlockjc replied:
Agreed, my bad.

misterdjules commented on test/simple/test-domain-with-abort-on-uncaught-exception.js:
```diff
@@ -0,0 +1,229 @@
+    fs.exists('/non/existing/file', function onExists(exists) {
+      throw new Error("Error from fs.exists callback");
+    });
+
```
I would also throw from within a `setTimeout`'s callback, as these are handled in a different code path as the others, and thus handle domains specifically too.

Something like:

```
setTimeout(function() { throw new Error("Error from setTimeout's callback"); }, 0);
```

whitlockjc replied:
You got it.

misterdjules commented:
@whitlockjc Added more comments.

Also, it would be great to split this PR in two different commits, the first one for the changes in deps/v8, and the second one for the changes outside of deps/v8. This way we can isolate and document this new floating patch in https://github.com/nodejs/node-v0.x-archive/wiki/V8-upgrade-process#v012x.

When these are addressed we can land this PR in v0.12. Thank you for your work again! :+1:

whitlockjc commented:
You got it.  I'll get on it.

whitlockjc commented:
Since you've already got a PR for this *(https://github.com/nodejs/node/pull/3036)*, how do you want me to handle this?  Create a new PR with two separate commits and that will replace the aforementioned PR?

misterdjules commented:
@whitlockjc We still don't know if we'll want to go with nodejs/node#3036 or nodejs/node#3038, or something else. So I would suggest just updating this PR.

whitlockjc commented:
Okay.  Thanks for the guidance Julien.

whitlockjc commented:
I've addressed the changes @misterdjules suggested.  I will not break the singular port commit into two port commits.  When this is complete, I will update you.

whitlockjc committed:
v8: port fbff705 from v0.10 to v0.12

fbff705
Add v8::Isolate::SetAbortOnUncaughtException() so the user can be
notified when an uncaught exception has bubbled.

whitlockjc committed:
domains: port caeb677 from v0.10 to v0.12

caeb677
Do not abort the process if an error is thrown from within a domain,
an error handler is setup for the domain and
--abort-on-uncaught-exception was passed on the command line.

However, if an error is thrown from within the top-level domain's
error handler and --abort-on-uncaught-exception was passed on the
command line, make the process abort.

Fixes: #8877

whitlockjc commented:
I have broken the PR commit into two commits, one for `v8` and one for `domains`.  I retained the original commit messages in their respective port commits but made a few minor changes to the subject to fit.

misterdjules commented:
LGTM. Thank you once again @whitlockjc!

misterdjules commented:
Landing with https://ci.nodejs.org/job/node-accept-pull-request/129/.

misterdjules commented:
Landed in 1982ed6e6356070241d695f77888a7681fd3c23c and f0453caea2c47405625236cf42ae11ae2a857142.

whitlockjc commented:
Is there a next step for getting this into nodejs/node?

misterdjules commented:
@whitlockjc These two commits are in nodejs/node's v0.12 branch. I'll follow-up on https://github.com/nodejs/node/pull/3036 when I get some feedback about the V8 changes in https://codereview.chromium.org/1375933003/. If that takes too long, or if it turns out https://github.com/nodejs/node/pull/3036 is not the solution we want to consider, we'll consider the other options mentioned in https://github.com/nodejs/node/pull/3035.

whitlockjc commented:
Gotcha.  Thanks for the heads up and let me know if I can help further regardless of the decisions mentioned above.
